bioimageio/core/prediction_pipeline/_processing.py

```diff
@@ -0,0 +1,228 @@
+from dataclasses import InitVar, dataclass, field
+from itertools import chain
+from typing import (
+    Callable,
+    ClassVar,
+    Dict,
+    FrozenSet,
+    List,
+    Literal,
+    Optional,
+    Sequence,
+    Set,
+    Tuple,
+    Type,
+    Union,
+    get_args,
+)
+
+import numpy as np
+import xarray as xr
+
+from bioimageio.core.resource_io import nodes
+from ._types import Transform
+from .statistical_measures import Mean, Measure, Percentile, Std
+from bioimageio.spec.model.v0_3.raw_nodes import PreprocessingName
+
+
+def ensure_dtype(tensor: xr.DataArray, *, dtype) -> xr.DataArray:
+    """
+    Convert array to a given datatype
+    """
+    return tensor.astype(dtype)
+
+
+@dataclass
+class Processing:
+    apply_to: str
+    computed_statistics: Dict[str, Dict[str, Measure]]
+
+    def get_required_statistics(self) -> Dict[str, Dict[str, Measure]]:
+        """
+        specifies which dataset measures are required for which purpose from what tensor.
+        Returns: dataset measures required to apply this processing indexed by <tensor_name> and <role>:
+                 Dict[<tensor_name>, Dict[<role>, Measure]]
+
+        """
+        return {}
+
+    def set_computed_statistics(self, computed: Dict[str, Dict[str, Measure]]):
+        """helper to set computed statistics and check if they match the requirements"""
+        for tensor_measures in computed.values():
+            for measure in tensor_measures.values():
+                assert measure.value is not None, "encountered uncomputed measure"
+
+        for tensor_name, req_measures in self.get_required_statistics():
+            comp_measures = computed.get(tensor_name, {})
+            for name, req_measure in req_measures.items():
+                comp_measure = comp_measures.get(name)
+                assert isinstance(comp_measure, type(req_measure))
+
+        self.computed_statistics = computed
+
+    def get_computed_statistics(self, tensor_name: str, measure_name: str):
+        """helper to unpack self.computed_statistics"""
+        ret = self.computed_statistics.get(tensor_name, {}).get(measure_name)
+        if ret is None:
+            raise RuntimeError(f"Missing {measure_name} for {tensor_name} dataset.")
+
+        return ret
+
+    def apply(self, **tensors: xr.DataArray) -> Dict[str, xr.DataArray]:
+        """apply processing to named tensors; call 'apply_simple' as default"""
+        tensors[self.apply_to] = self.apply_simple(tensors[self.apply_to])
+        return tensors
+
+    def apply_simple(self, tensor: xr.DataArray) -> xr.DataArray:
+        """apply processing to single tensor"""
+        raise NotImplementedError
+
+    def __post_init__(self):
+        """validate common kwargs by their annotations"""
+        if hasattr(self, "mode"):
+            if self.mode not in get_args(self.mode):
+                raise NotImplementedError(f"Unsupported mode {self.mode} for {self.__class__.__name__}: {self.mode}")
```

Comment:
This is basically the whole sketch: with this base we can hide all the tensor name and stats complexity for simple cases and have convenient access to them when we need it...